lang/udf_cache.hh

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "cql3/functions/function_name.hh"
+#include "utils/overloaded_functor.hh"
+#include "wasmtime.hh"
+#include "lang/wasm.hh"
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/scheduling.hh>
+#include "utils/loading_cache.hh"
+#include "utils/hash.hh"
+#include "utils/overloaded_functor.hh"
+#include <chrono>
+#include <cstdint>
+#include <queue>
+#include <unordered_map>
+
+extern logging::logger wasm_logger;
+
+namespace wasm {
+
+struct udf_instance {
+    wasmtime::Store store;
+    wasmtime::Instance instance;
+    wasmtime::Func func;
+    wasmtime::Memory memory;
+};
+
+// For each UDF full name and a scheduling group, we store a future that is ready
+// when no other task uses the same UDF instance. When an instance is retrieved
+// from the cache, the future is replaced with a new one that becomes ready when
+// the instance is returned. To avoid conflicts between tasks, each time an 
+// instance is requested, the future associated with it is removed from the cache
+// and a new one is added. When the current instance user is done with it, the
+// future that was associated with it is becomes ready.
+// This way, an instance stays in the cache as long as it isn't used by any task
+// and if there are tasks that want to use it, they get it immediately after
+// the previous user is finished.
+class udf_cache {
+public:
+    struct stats {
+        uint64_t cache_hits = 0;
+        uint64_t cache_misses = 0;
+        uint64_t cache_blocks = 0;
+    };
+
+    static stats& shard_stats() {
+        static thread_local stats _stats;
+        return _stats;
+    }
+
+private:
+    using cache_key_type = std::pair<seastar::scheduling_group, cql3::functions::function_name>;
+    struct cache_entry_type {
+        mutable udf_instance inst;
+    };
+    struct udf_cache_entry_size {
+        size_t operator()(cache_entry_type& val) {
+            return val.inst.memory.size(val.inst.store);
+        }
+    };
+    using cache_type = utils::loading_cache<cache_key_type, cache_entry_type, 1, utils::loading_cache_reload_enabled::no, udf_cache_entry_size, utils::tuple_hash, std::equal_to<cache_key_type>>;
+
+public:
+    using key_type = cql3::functions::function_name;
+    using value_type = udf_instance;
+
+private:
+    std::unordered_map<cache_key_type, std::queue<seastar::promise<value_type>>, utils::tuple_hash> _in_use_instances;
```

Comment:
multimap?

Reply:
This is not the cache that keeps multiple instances for a single function name. Using this map we plan how the instances outside the cache are going to be reused (we arrange them in a queue)

Reply:
Why have a special case for that? Put those instances back in the cache, and have the waiters take them from the cache. Isn't it simpler?

Reply:
There is no distinction in the next version, but I see that it would work if we did have no values in the promises/futures and only in the cache. It also looks like the `seastar::shared_mutex` is almost exactly that (queue of promises with no values)